armi/physics/neutronics/crossSectionGroupManager.py

```diff
@@ -387,6 +387,119 @@ def getNumberDensityWithTrace(component, nucName):
+            allNucsNames, densities = self._getAverageComponentNucs(
+                allSimilarComponents, bWeights
+            )
+            for nuc, aDensity in zip(allNucsNames, densities):
+                c.setNumberDensity(nuc, aDensity)
+        return repBlock
+
+    def _getNucTempHelper(self):
+        raise NotImplementedError
+
+    @staticmethod
+    def _getAllNucs(components):
+        """Iterate through components and get all unique nuclides."""
+        nucs = set()
+        for c in components:
+            nucs = nucs.union(c.getNuclides())
+        return sorted(list(nucs))
+
+    @staticmethod
+    def _checkComponentConsisteny(b, repBlock):
+        """
+        Verify that all components being homogenized are rectangular and have consistent dimensions.
+
+        Raises
+        ------
+        ValueError
+            When the components in a candidate block do not align with
+            the components in the representative block. This check includes component area, component multiplicity,
+            and nuclide composition.
+        """
+        for c, repC in zip(b, repBlock):
+            compString = (
+                f"Component {repC} in block {repBlock} and component {c} in block {b}"
+            )
+            if c.p.mult != repC.p.mult:
+                raise ValueError(
+                    f"{compString} must have the same multiplicity, but they have."
+                    f"{repC.p.mult} and {c.p.mult}, respectively."
+                )
+
+            theseNucs = set(c.getNuclides())
+            thoseNucs = set(repC.getNuclides())
+            diffNucs = theseNucs.symmetric_difference(thoseNucs)
+            if diffNucs:
+                raise ValueError(
+                    f"{compString} are in the same location, but nuclides "
+                    f"differ by {diffNucs}. \n{theseNucs} \n{thoseNucs}"
+                )
+
+    def _getAverageComponentNucs(self, components, bWeights):
+        """Compute average nuclide densities by block weights and component area fractions."""
+        allNucNames = self._getAllNucs(components)
+        densities = numpy.zeros(len(allNucNames))
+        totalWeight = 0.0
+        for c, bWeight in zip(components, bWeights):
+            weight = bWeight * c.getArea()
+            totalWeight += weight
+            densities += weight * numpy.array(c.getNuclideNumberDensities(allNucNames))
+        return allNucNames, densities / totalWeight
+
+    def _orderComponentsInGroup(self, repBlock):
+        """Order the components based on dimension and material type within the representative block."""
+        orderedComponents = [[] for _ in repBlock]
+        for b in self.getCandidateBlocks():
+            if len(b) != len(repBlock):
+                raise ValueError(
+                    f"Blocks {b} and {repBlock} have differing number "
+                    f"of components and cannot be homogenized"
+                )
```

Comment:
This could probably be moved to within `_checkComponentConsistency`

Reply:
The whole thing then could possibly be rewritten into:

```
for b in self.getCandidateBlocks():
    self._checkComponentConsistency(b, repBlock)
componentLists = [ list(b) for b in self.getCandidateBlocks() ]
return [ list(comps) for comps in zip(*componentLists) ]
```

(just a suggestion here)